### PR TITLE
Promote arm64 macOS to tier 1

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Github recently added a new macOS runner that is using the m1 CPU that is usable for open source projects. [1] Previously Qiskit had support for arm64 macOS at tier 3 because we were only able to cross compile for the platform and not test the binaries. Now that we can run CI jobs on the platform we're able to run both unit tests and test our binaries on release. This commit adds a new set of test jobs and wheel builds that use the macos-14 runner that mirrors the existing x86_64 macOS jobs we have. This brings arm64 macOS to the same support level as x86_64. THe only difference here is that azure pipelines doesn't have arm64 macOS support so the test job will run in github actions instead.

### Details and comments

[1] https://github.blog/changelog/2024-01-30-github-actions-introducing-the-new-m1-macos-runner-available-to-open-source/

### CODE REVIEW

**Comment by mtreinish** [comment]
I've marked this as on hold until we verify the job configurations work.

**Comment by mtreinish** [comment]
So there are 2 big issues so far with this. The first is 6 to 12 tests are failing because of numpy precision issues that probably come down to SIMD differences or issues with accelerate on arm64. This also causes the PGO test runs to fail in the wheel builds.

The second issue is that there are no Python 3.8 official Python binaries available on macOS arm64. We were previously publishing arm64 py3.8 wheels with cross compilation on x86_64, but now that we're doing native builds for arm64 the builds fail because it tries to run with x86_64 python. I think the only path to workaround that is to add back the cross-build job and have it only run for python 3.8. This precludes doing PGO, but I think the number of arm64 macOS Python 3.8 users will be very low if there aren't official Python packages for it so that's an ok tradeoff.

**Comment by mtreinish** [comment]
I've pushed up #12197 to remove linear algebra with numpy from the equation in the isometry code path to fix the issues we're having here. This should be good to go once #12197 merges and we update the PR branch

I've tested that tests work here: https://github.com/mtreinish/qiskit-core/actions/runs/8730127221 and I'm debugging the wheel builds here: https://github.com/mtreinish/qiskit-core/actions/runs/8730127218/job/23953358067 once that's working I'll remove the test commits from this branch to get this in a mergable state again.

The only other change I think we should make here is tweaking one of the test jobs to run with MSRV instead of the latest stable. But that's trivial to do.

**Comment by mtreinish** [comment]
I've tested that unittest jobs: https://github.com/mtreinish/qiskit-core/actions/runs/8736818119

and wheel builds: https://github.com/mtreinish/qiskit-core/actions/runs/8736818125

with the current state of this PR and #12197 applied. So I'm going to revert https://github.com/Qiskit/qiskit/pull/12102/commits/97eaa6fee84d8209e530e59017478c78bce1a868 and prepare this final merging now. The test CI jobs will still fail until we merge #12197 but otherwise I think this will be ready.

File: .github/workflows/wheels.yml
Comment (kevinhartman) [inline .github/workflows/wheels.yml]: I think it's worth adding a comment (or even changing the title) for this job, since it's only here to build the macOS Python 3.8 wheel for ARM, correct?

**Review by kevinhartman — COMMENTED** [review]
Looks mostly good.

It's really cool to see us finally able to run CI for macOS ARM variants :D.

File: .github/workflows/wheels.yml
Comment (mtreinish) [reply .github/workflows/wheels.yml]: Sure, I'll update the job name

Comment (mtreinish) [reply .github/workflows/wheels.yml]: Done in: https://github.com/Qiskit/qiskit/pull/12102/commits/a340b51b1be6e8727ec1ba1222a7764735142c10

**Review by kevinhartman — APPROVED** [review]

**Pull request merged**
